### PR TITLE
Page numbers header footer

### PR DESCRIPTION
**Pull request by benjaminketron**
`header` and `footer` functionality for caracal as well as paragraph `page` and `numpages` helper methods.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -317,6 +317,35 @@ end
 The `size` option and the `label_size` and `number_size` options are mutually exclusive.
 
 
+### Footer
+
+A footer appearing on every page can be added via the `footer` method. Its block takes a footer object that supports text and table methods.
+
+*This is superceded by `page_numbers`, but can be used instead to provide a more complext page number footer.*
```
Comment (jrforrest) [inline README.md]: Typo: `complext`

File: lib/caracal/core/footer.rb [hunk 2]
```diff
@@ -0,0 +1,34 @@
+require 'caracal/core/models/footer_model'
+require 'caracal/errors'
+
+
+module Caracal
+  module Core
+
+    # This module encapsulates all the functionality related to adding tables
```
Comment (jrforrest) [inline lib/caracal/core/footer.rb]: Bad copypasta on this doc?

File: lib/caracal/core/header.rb [hunk 3]
```diff
@@ -0,0 +1,32 @@
+require 'caracal/core/models/header_model'
+require 'caracal/errors'
+
+
+module Caracal
+  module Core
+
+    # This module encapsulates all the functionality related to adding tables
```
Comment (jrforrest) [inline lib/caracal/core/header.rb]: Bad doc here as well?

File: lib/caracal/core/relationships.rb [hunk 4]
```diff
@@ -4,42 +4,43 @@
...
           #-------------------------------------------------------------
           # Class Methods
           #-------------------------------------------------------------
-          
+
           def self.default_relationships
             [
               { target: 'fontTable.xml',  type: :font      },
               { target: 'footer1.xml',    type: :footer    },
+              { target: 'header1.xml',    type: :header    },
```
Comment (jrforrest) [inline lib/caracal/core/relationships.rb:30]: Why `header1` rather than just `header.xml`?

File: lib/caracal/renderers/header_renderer.rb [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+module Caracal
+  module Renderers
+    class HeaderRenderer < DocumentRenderer
+
+      #-------------------------------------------------------------
+      # Public Methods
+      #-------------------------------------------------------------
+
+      # This method produces the xml required for the `word/settings.xml`
```
Comment (jrforrest) [inline lib/caracal/renderers/header_renderer.rb]: This comment isn't valid for this file, right?  This produces xml require for `word/header1.xml`?

**Review by jrforrest — CHANGES_REQUESTED** [review]
Whew this coding style is really verbose!  I think you did the right thing in keeping with the upstream projects styling though.  I did recommend a few minor changes here.

Also, I see that there are some commits on this PR for the v1.4.1 version bump.  Should those really be included here?  We'll be referencing this dependency on our projects via a git-ref, and I don't think an upstream PR should include the version bump, right?

File: README.md (hunk 1 above)
Comment (benjaminketron) [reply README.md]: 🙃

File: lib/caracal/core/footer.rb (hunk 2 above)
Comment (benjaminketron) [reply lib/caracal/core/footer.rb]: oops. yeah. thanks for catching that.

File: lib/caracal/core/relationships.rb (hunk 4 above)
Comment (benjaminketron) [reply lib/caracal/core/relationships.rb:30]: Was following the footer convention. Not sure if there is a reason to enumerate the header filename other than there could technically be multiple.

File: lib/caracal/renderers/header_renderer.rb (hunk 5 above)
Comment (benjaminketron) [reply lib/caracal/renderers/header_renderer.rb]: Ah, thanks for catching that! Looks like there is a similar issue with the footer renderer.

File: lib/caracal/core/header.rb (hunk 3 above)
Comment (benjaminketron) [reply lib/caracal/core/header.rb]: yeah. thanks for catching that!

File: lib/caracal/core/relationships.rb (hunk 4 above)
Comment (benjaminketron) [reply lib/caracal/core/relationships.rb:30]: Tempted to keep it since it is alignment with the footer file name convention.

**Comment by benjaminketron** [comment]
@jrforrest Thanks for the review! A rebase with master appears to have solved the version bump issue. I'd half noticed it but didn't really give it much thought until you called it out. Not 100% how that happened.... must have branched from a  previous commit. :)

File: lib/caracal/core/relationships.rb (hunk 4 above)
Comment (jrforrest) [reply lib/caracal/core/relationships.rb:30]: Sure, not a big difference, just a tad goofy :)

Comment (benjaminketron) [reply lib/caracal/core/relationships.rb:30]: 🙃

**Review by jrforrest — APPROVED** [review]
Awesome.  Good cleanup Ben I think we're good to go with merging this guy!

**Comment by benjaminketron** [comment]
@jrforrest Thanks!

**Pull request merged**